terraso_backend/apps/project_management/permission_rules.py

```diff
@@ -12,137 +12,118 @@
+from apps.project_management.models import Project, Site, SiteNote
 
 
-@rules.predicate
-def allowed_to_delete_site(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user)
+@dataclass
+class Context:
+    # these are the target entity(s) the operation is occurring on. they will always be
+    # referentually consistent (i.e., if the context has a site this will always be the site's
+    # project)
+    project: Project = None
+    site: Site = None
+    site_note: SiteNote = None
 
+    # these are the source entities which may also be part of the operation (i.e., the site which
+    # is being added to a new owner, and that site's current project if it has one)
+    source_project: Project = None
+    source_site: Site = None
 
-@rules.predicate
-def allowed_to_update_site_settings(user, site):
-    if site.owned_by_user:
-        return site.owner == user
-    return site.project.is_manager(user)
-
-
-@rules.predicate
-def allowed_to_delete_project(user, project):
-    return project.is_manager(user)
+    def __post_init__(self):
+        if self.site_note:
+            self.site = self.site_note.site
+        if self.site:
+            self.project = self.site.project
+        if self.source_site:
+            self.source_project = self.source_site.project
 
 
 @rules.predicate
-def allowed_to_add_to_project(user, project):
-    return project.is_manager(user)
+def allowed_to_create(user, context):
+    # (all logged-in users are allowed base creation actions)
+    return True
 
 
 @rules.predicate
-def allowed_to_archive_project(user, project):
-    return project.is_manager(user)
+def allowed_to_manage_project(user, context):
+    return context.project.is_manager(user)
 
 
 @rules.predicate
-def allowed_to_add_member_to_project(user, context):
-    project = context["project"]
-    requester_membership = context["requester_membership"]
+def allowed_to_be_project_member(user, context):
     return (
-        requester_membership.membership_list == project.membership_list
-        and requester_membership.user_role == ProjectRole.MANAGER.value
+        context.project.is_manager(user)
+        or context.project.is_contributor(user)
+        or context.project.is_viewer(user)
     )
 
 
-rules.add_rule("allowed_to_add_member_to_project", allowed_to_add_member_to_project)
-
-
 @rules.predicate
-def allowed_to_delete_user_from_project(user, context):
-    project = context["project"]
-    requester_membership = context["requester_membership"]
-    target_membership = context["target_membership"]
-    return project.membership_list == requester_membership.membership_list and (
-        user == target_membership.user
-        or requester_membership.user_role == ProjectRole.MANAGER.value
-    )
+def allowed_to_contribute_to_affiliated_site(user, context):
+    if context.site.is_unaffiliated:
+        return False
```

Comment:
question: how does this work? why do we want to return false if the site is unaffiliated? Same check is used through the file.

Reply:
If a site is unaffiliated, it definitionally cannot have any "affiliated"-type actions applied to it because there is no affiliated project.

Reply:
in that case, should this be an assertion or throw an error? it seems like more of a programming error than a permissions failure to me if allowed_to_X_affiliated_site is called with an unaffiliated site